projects/graph/graph.py

```diff
@@ -3,56 +3,153 @@
         This should be done using recursion.
         """
-        pass  # TODO
+        if visited is None:
+            visited = set()
+        if starting_vertex not in visited:
+            print(starting_vertex)
+            visited.add(starting_vertex)
+            for neighbor in self.vertices[starting_vertex]:
+                self.dft_recursive(neighbor, visited)
+
     def bfs(self, starting_vertex, destination_vertex):
         """
         Return a list containing the shortest path from
         starting_vertex to destination_vertex in
         breath-first order.
         """
-        pass  # TODO
+        # Create an empty set to store visited nodes
+        visited = set()
+        # Create an empty Queue and enqueue A PATH TO the starting vertex
+        q = Queue()
+        q.enqueue([starting_vertex])
+        # While the queue is not empty...
+        while q.size() > 0:
+            # Dequeue the first PATH
+            path = q.dequeue()
+        # GRAB THE VERTEX FROM THE END OF THE PATH
+            v = path[-1]
+        # IF VERTEX == TARGET, RETURN PATH
+            if v == destination_vertex:
+                return path
+        # If that vertex has not been visited...
+            if v not in visited:
+                # Mark it as visited
+                visited.add(v)
+        # Then add A PATH TO all of its neighbors to the back of the queue
+                for neighbor in self.vertices[v]:
+                    # Copy the path
+                    path_copy = list(path)
+        # Append neighbor to the back of the copy
+                    path_copy.append(neighbor)
+        # Enqueue copy
+                    q.enqueue(path_copy)
+
     def dfs(self, starting_vertex, destination_vertex):
         """
         Return a list containing a path from
         starting_vertex to destination_vertex in
         depth-first order.
         """
-        pass  # TODO
-
-
-
+        # Create an empty set to store visited nodes
+        visited = set()
+        # Create an empty Stack and push A PATH TO the starting vertex
+        s = Stack()
+        s.push([starting_vertex])
+        # While the stack is not empty...
+        while s.size() > 0:
+            # Pop the first PATH
+            path = s.pop()
+        # GRAB THE VERTEX FROM THE END OF THE PATH
+            v = path[-1]
+        # IF VERTEX == TARGET, RETURN PATH
+            if v == destination_vertex:
+                return path
+        # If that vertex has not been visited...
+            if v not in visited:
+                # Mark it as visited
+                visited.add(v)
+        # Then add A PATH TO all of its neighbors to the top of the stack
+                for neighbor in self.vertices[v]:
+                    # Copy the path
+                    path_copy = list(path)
+        # Append neighbor to the back of the copy
+                    path_copy.append(neighbor)
+        # Enqueue copy
+                    s.push(path_copy)
 
```

Comment:
Nice work with BFS/DFS methods.  Isnt it crazy how similar the code for them is?  In the future i bet we could make a function that handles both cases based on a passed-in parameter because of how similar the code.

I also like how the code is well commented.